kythe/go/util/span/span_test.go

```diff
@@ -157,9 +158,150 @@ func TestPatcher(t *testing.T) {
 	}
 }
 
+func p(bo, ln, co int32) *cpb.Point {
+	return &cpb.Point{ByteOffset: bo, LineNumber: ln, ColumnOffset: co}
+}
+
+func sp(start, end *cpb.Point) *cpb.Span { return &cpb.Span{Start: start, End: end} }
+
+func TestPatchSpan(t *testing.T) {
+	tests := []struct {
+		oldText, newText string
+
+		oldSpans []*cpb.Span
+		newSpans []*cpb.Span
+	}{{
+		oldText: "this is some text",
+		newText: "this is some changed text",
+
+		oldSpans: []*cpb.Span{
+			sp(p(0, 1, 0), p(4, 1, 4)),
+			sp(p(13, 1, 13), p(17, 1, 17)),
+		},
+		newSpans: []*cpb.Span{
+			sp(p(0, 1, 0), p(4, 1, 4)),
+			sp(p(21, 1, 21), p(25, 1, 25)),
+		},
+	}, {
+		oldText: "line one\nline two\nline three\nline four\n",
+		newText: "line one\nline three\nline two\npre line four\n",
+
+		oldSpans: []*cpb.Span{
+			sp(p(0, 1, 0), p(5, 1, 5)),
+			sp(p(10, 2, 0), p(14, 2, 4)),
+			sp(p(15, 2, 6), p(18, 2, 11)),
+			sp(p(30, 4, 0), p(34, 4, 4)),
+		},
+		newSpans: []*cpb.Span{
+			sp(p(0, 1, 0), p(5, 1, 5)),
+			sp(p(10, 2, 0), p(14, 2, 4)),
+			nil,
+			sp(p(34, 4, 4), p(38, 4, 8)),
+		},
+	}, {
+		oldText: "line one\nmoved\n",
+		newText: "line one\ninsert\nmoved\n",
+
+		oldSpans: []*cpb.Span{
+			sp(p(5, 1, 5), p(8, 1, 8)),
+			sp(p(10, 2, 0), p(15, 2, 5)),
+		},
+		newSpans: []*cpb.Span{
+			sp(p(5, 1, 5), p(8, 1, 8)),
+			sp(p(17, 3, 0), p(22, 3, 5)),
+		},
+	}, {
+		oldText: "line\ntwo\nthree\n",
+		newText: "line one\nline two\nthee\n",
+
+		oldSpans: []*cpb.Span{
+			sp(p(0, 1, 0), p(4, 1, 4)),
+			sp(p(5, 2, 0), p(8, 2, 3)),
+			sp(p(9, 3, 0), p(14, 3, 5)),
+		},
+		newSpans: []*cpb.Span{
+			sp(p(0, 1, 0), p(4, 1, 4)),
+			sp(p(14, 2, 5), p(17, 2, 8)),
+			nil,
+		},
+	}, {
+		oldText: "line three\n",
+		newText: "line one\ntwo\nthree\n",
+
+		oldSpans: []*cpb.Span{
+			sp(p(0, 1, 0), p(4, 1, 4)),
+			sp(p(5, 1, 5), p(10, 1, 10)),
+		},
+		newSpans: []*cpb.Span{
+			sp(p(0, 1, 0), p(4, 1, 4)),
+			sp(p(13, 3, 0), p(18, 3, 5)),
+		},
+	}, {
+		oldText: "line three\nfour\n",
+		newText: "line one\ntwo\nthree\nfour\n",
+
+		oldSpans: []*cpb.Span{
+			sp(p(0, 1, 0), p(4, 1, 4)),
+			sp(p(5, 1, 5), p(10, 1, 10)),
+			sp(p(11, 2, 0), p(15, 2, 4)),
+		},
+		newSpans: []*cpb.Span{
+			sp(p(0, 1, 0), p(4, 1, 4)),
+			sp(p(13, 3, 0), p(18, 3, 5)),
+			sp(p(19, 4, 0), p(23, 4, 4)),
+		},
+	}, {
+		oldText: "線\n動\n",
+		newText: "線\n入\n動\n",
```

Comment:
include a test without a final newline?

Reply:
The 0th test doesn't have a trailing newline.